### DISCUSSION TITLE
Auto-matting only *some* images

### DISCUSSION THREAD

**Discussion by jgodfrey** [helgeerbe/picframe#38] [Ideas]
Originally, the auto-matting feature was either _ON_ (all images were matted) or _OFF_ (no images were matted). Helge has since added an experimental `mat_portraits_only` feature that's designed to auto-mat only portrait images.

(As we've discussed elsewhere), I think we need a more general way to control the matting of only _some_ images - likely based on comparing the aspect ratio of the image to that of the screen. For example, we could auto-mat images whose aspect ratio doesn't match the screen's aspect ratio (within some specified tolerance). 

Assuming we add such a feature (and I think we should), does that eliminate the need for Helge's explicit _`mat_portraits_only` addition? I think it does, as it'd allow images that don't fit the screen to be auto-matted, which I think is the main motivation behind Helge's addition.

Also, it seems a bit more flexible as it'd allow landscape images (that didn't fit the screen aspect) to be auto-matted too.

Of course, all of this would require that the `mat_images` property was set to _True_ to start with. That is, you don't have to mat anything if you don't want to.

Really, just trying to decide whether this new option needs to coexist with the `mat_portraits_only` feature, or whether it can _replace_ it...

**Comment by paddywwoof** [comment]
My initial feeling is that ```matting_non_display_shaped_images``` trumps ```mat_portraits_only``` in pretty much all circumstances. So for the sake of simplicity I would suggest opting for a replacement of the code and the config option. (With a punchier name)

**Reply by jgodfrey** [reply]
> My initial feeling is that `matting_non_display_shaped_images` trumps `mat_portraits_only` in pretty much all circumstances.

Yep, that's my feeling as well. For configuration, I'm contemplating using the single, existing `mat_images` property (which currently is just a bool) and extending it to something like:

`True` - mat all images
`False` - mat no images
`Some Real Value` - the aspect ratio tolerance for matting.

That'd require the value to be parsed and error checked in the code, but I wonder if that's not easier for an end user to handle, rather than interacting with separate properties. 

Any thoughts?

**Reply by jgodfrey** [reply]
Really, the matting of all, none, or some images could be handled with a single, real value representing the tolerance. For example, a sufficiently small number (maybe even -1) would mat everything, a sufficiently large number would mat nothing, and something in between would mat only "some" images.

While that's easier to work with in the code, I think it's far less obvious to an end user for the _all_ or _nothing_ cases than the above True/False/Real option outlined above.

**Reply by sapnho** [reply]
> True - mat all images
False - mat no images
Some Real Value - the aspect ratio tolerance for matting.

Making it really easy for the user will determine if people are going to use it. Yes and No is easy enough. The aspect ratio tolerance is probably too much math for most people. I would just assume a value here. The other thing is portrait pairs yes/no.

**Reply by jgodfrey** [reply]
> Making it really easy for the user will determine if people are going to use it.

I agree. Though, we need to decide what actually _is_ easiest for the user. Maybe the place to start would be: "What do we think the default state of the auto-matting system should be?" I'd guess it should be to "automatically mat images that don't fit the screen aspect".

If that's the case, we could still go with my above suggestion of a single property with "true / false / real" values. The above-described default state would require the property to contain an appropriate tolerance value (so, the "REAL" option). We could just supply that directly in the default config file. 

That way, a new User would get auto-matting of non-aspect-matching images by default. If they wanted to change that (to either mat everything or mat nothing), they'd only have to supply the "more standard" boolean values to the config property. So, `True` to mat everything or `False` to mat nothing.

It still seems to me that controlling the use of the matting system via a single property is the simplest for a user. Though, I agree, expecting them to understand (and properly supply) some tolerance value is probably a bridge too far. So, I suggest that we supply a suitable tolerance value as the default and let them update the value with a simple `True` / `False` if they want different behavior.

And, the _portrait_pairs_ is an independent feature now, which works correctly with or without the auto-matting. I assume that can stay as-is, right?

**Reply by sapnho** [reply]
I will write up a full article on the new matting feature where I can walk people through the various options. I am not worried about making the options too complicated in the config but there should be a very simple default setting that most people can just accept.

> I'd guess it should be to "automatically mat images that don't fit the screen aspect".

That sounds good to me. And actually, below that ON/OFF option, you can list all the other options, I can explain everything in the article and give examples.

PictureFrame also has many options that even took me a while to understand and only when I wrote the article on all the Pi3D options and tested every single one, I eventually understood.

**Reply by paddywwoof** [reply]
And even  now I get confused about "fit": whether it stretches the images so it fits but some gets cropped or whether is shrinks the image so it fits but leaves a gap (I know Wolfgang and Helge just make all their images the same size as the screen!). We should probably change the name of that ```whole_picture_leave_gap:False``` or ```crop_picture_no_gap:True```

This is probably relevant to the threshold for matting. If the user chooses ```crop_picture...``` then they probably don't mind having 5% sliced off the top and bottom but if they choose ```whole_picture...``` they might be annoyed by 1% of blur or reflected image

**Reply by sapnho** [reply]
Oh yes, the "Fit" parameter is indeed very confusing, but that is the case with the commercial photo frames as well. What the heck is "Fit to screen"? :-) Some call it "Fill screen" but do not specify that this entails cropping.
So what would be terminology that everybody can understand? Tricky..

**Reply by jgodfrey** [reply]
On the auto matting config mechanism, I have some code working as described above. Specifically...

- If `mat_images` is `True`, all images are matted.
- If `mat_images` is `False`, no images are matted.
- If `mat_images` is a valid real value, only images whose aspect is mismatched from the screen by more than that value are matted.

We'll need to decide what the default config value should be, but I'm currently running with `0.01`. So, images off by more than 1% will be matted.

Behind the scenes, I'm turning the single `mat_images` config value into 2 values. One value indicates whether the auto-mat feature is ON or OFF and the other value is the aspect tolerance (assuming the feature is on). That's done like:

- If the config value is (some form of) `True`, the features is on and the tolerance is set to `-1`. This will mat everything.
- If the config value is (some form of) `False`, the feature is off. In this case, the tolerance value doesn't matter.
- If the config value is a valid real number, the features is on and the value is the tolerance.
- If the config value is none of the above (so, garbage), I issue a log warning and use the defaults (feature on, 0.01 tolerance).

Any objections?

**Reply by sapnho** [reply]
I guess, we'll have to see how it works in real life but I believe you have tested it with different aspect ratios already, have you? If not, I can include a fat "Give the author feedback" button in my article and we'll find out quickly how it turns out. :-)

**Reply by jgodfrey** [reply]
> I believe you have tested it with different aspect ratios already, have you?

Certainly. I'm testing with a completely uncurated set of images. So, no special cropping, some "free" cropping, multiple orientations, and varying devices. So, probably dozens of different aspect ratios.

**Comment by sapnho** [comment]
Hi @jgodfrey, just posted my article on the Matting feature: https://www.thedigitalpictureframe.com/activate-the-power-of-the-magic-matting-feature-on-your-raspberry-pi-picture-frame/

I noticed that in our default configuration.yaml file the styles are not exactly the same as in the Wiki  and we should probably align the two.

> "A string containing the mat types to choose from when matting images. It can consist of any or all of 'float float_polaroid single_bevel double_bevel double_flat' (null or '' will use all mat types)"

Matting is a superb new feature of PictureFrame. Thank you!

**Reply by jgodfrey** [reply]
Really nice article Wolfgang - well done! Regarding the matting styles definition being incorrect in the default config file, I think that should be fixed with this commit from 9 days ago. I assume it just hasn't made it into packaged release...

https://github.com/helgeerbe/picframe/commit/7b16e4ea6b7632c274b905bd85c4a46b6ec755ff

Regarding this statement from your article:

_Currently, there are six different styles (“mat-type”) included. You can either specify your favorite style or use all of them at random._

It doesn't have to be either _all_ or _one_. You can also simply list the styles you want to use. So, for example, if you don't like 2 of the styles, you can set `mat_type` to a string that contains the 4 styles you do like. The styles listed in the string will be chosen from at random during matting - whether that's one style, all styles, or somewhere in between...

**Reply by jgodfrey** [reply]
> I assume it just hasn't made it into packaged release...

After getting through all of my email this morning, I see this was released today.

**Reply by sapnho** [reply]
> It doesn't have to be either _all_ or _one_. You can also simply list the styles you want to use. So, for example, if you don't like 2 of the styles, you can set `mat_type` to a string that contains the 4 styles you do like. The styles listed in the string will be chosen from at random during matting - whether that's one style, all styles, or somewhere in between...

How would that look? Because when I enter "mat_type: "float float_polaroid"" it only shows the polaroid style.

**Reply by jgodfrey** [reply]
> How would that look? Because when I enter "mat_type: "float float_polaroid"" it only shows the polaroid style.

Exactly that should work (or separated by commas). Are you sure it doesn't? Let it run a little longer, as it may repeat one of the styles a few times. Also, make sure you don't have a typo in a style name as it'll simply be ignored (though, it'll log a DEBUG level message for any style names it doesn't recognize).

**Reply by sapnho** [reply]
Hi @jgodfrey, not sure what it may have been (me being impatient..) but it works as advertised now! Thanks! :-)
